src/app/dao/tests/PartDAO.spec.ts

```diff
@@ -60,43 +84,363 @@ describe('PartDAO', () => {
+            // Given a non-started part with its owner who has timed out
+            const creator: MinimalUser = await createConnectedUser(CREATOR_EMAIL, CREATOR_NAME);
+            const partId: string = await partDAO.create({ ...PartMocks.INITIAL, playerZero: creator });
+            await joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator });
+            // eslint-disable-next-line camelcase
+            const last_changed: Timestamp = new Timestamp(0, 0); // owner is stuck in 1970
+            // eslint-disable-next-line camelcase
+            await userDAO.update(creator.id, { observedPart: partId, last_changed });
+            await signOut();
+
+            // and given another user
+            await createConnectedUser(CANDIDATE_EMAIL, CANDIDATE_NAME);
+
+            // When the other user deletes the part
+            const result: Promise<void> = partDAO.delete(partId);
+
+            // Then it should succeed
+            await expectAsync(result).toBeResolvedTo();
+        });
+        it('should forbid deleting a non-started part if owner is still observing it and has not timed out', async() => {
+            // Given a part with its owner
+            const creator: MinimalUser = await createConnectedUser(CREATOR_EMAIL, CREATOR_NAME);
+            const partId: string = await partDAO.create({ ...PartMocks.INITIAL, playerZero: creator });
+            await joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator });
+            // eslint-disable-next-line camelcase
+            const last_changed: Timestamp = new Timestamp(Math.floor(Date.now() / 1000), 0);
+            // eslint-disable-next-line camelcase
+            await userDAO.update(creator.id, { observedPart: partId, last_changed });
+            await signOut();
+
+            // and given another user
+            await createConnectedUser(MALICIOUS_EMAIL, MALICIOUS_NAME);
+
+            // When the other user deletes the part
+            const result: Promise<void> = partDAO.delete(partId);
+
+            // Then it should fail
+            await expectPermissionToBeDenied(result);
+        });
+        it('should allow starting a part when chosen as opponent', async() => {
+            // Given a part, an accepted config, and a user who is the chosen opponent in the joiner
+            // Creator creates the joiner
+            const creator: MinimalUser = await createConnectedUser(CREATOR_EMAIL, CREATOR_NAME);
+            const part: Part = { ...PartMocks.INITIAL, playerZero: creator };
+            const partId: string = await partDAO.create(part);
+            await joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator });
+            await signOut();
+
+            // A candidate adds themself to the candidates list
+            const candidate: MinimalUser = await createConnectedUser(CANDIDATE_EMAIL, CANDIDATE_NAME);
+            await expectAsync(joinerDAO.addCandidate(partId, candidate)).toBeResolvedTo();
+            await signOut();
```

Comment:
that was sexy as a monoline function call like in JoinerDAO.spec.ts 🥉

Reply:
comme ceci ?